### ISSUE TITLE
BatchNorm2d permutes data on 1.10 nightly

### ISSUE THREAD

**Issue by david-macleod** [pytorch/pytorch#63224]
## 🐛 Bug

Passing a non-contiguous input to batchnorm2d causes the order of the output data to change. This doesn't happen on version 1.9.1


## To Reproduce

```py
import torch
print(torch.__version__)

x = torch.arange(6, dtype=torch.float).reshape(1,3,2,1)
bn = torch.nn.BatchNorm2d(2).eval()
x = x.permute(0,2,1,3)
print(x)
print(bn(x))
```
**PyTorch 1.9.1 (working)**
```
1.9.0+cu102
tensor([[[[0.],
          [2.],
          [4.]],

         [[1.],
          [3.],
          [5.]]]])
tensor([[[[0.0000],
          [2.0000],
          [4.0000]],

         [[1.0000],
          [3.0000],
          [5.0000]]]], grad_fn=<NativeBatchNormBackward>)
```

**PyTorch nightly 1.10 (broken)**
```
1.10.0.dev20210812+cu102
tensor([[[[0.],
          [2.],
          [4.]],

         [[1.],
          [3.],
          [5.]]]])
tensor([[[[0.0000],
          [1.0000],
          [2.0000]],

         [[3.0000],
          [4.0000],
          [5.0000]]]], grad_fn=<NativeBatchNormBackward>)
```

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->

## Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

## Environment

```
Collecting environment information...                                                                                                                                     
PyTorch version: 1.10.0.dev20210812+cu102                                                                                                                                 
Is debug build: False                                                                                                                                                     
CUDA used to build PyTorch: 10.2                                                                                                                                          
ROCM used to build PyTorch: N/A                                                                                                                                                                                                                                                                                                                     OS: Ubuntu 18.04.5 LTS (x86_64)                                                                                                                                           
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0                                                                                                                          Clang version: Could not collect                                                                                                                                          
CMake version: version 3.21.1                                                                                                                                            
Libc version: glibc-2.27                                                                                                                                                                                                                                                                                                                            Python version: 3.8.11 (default, Jul  3 2021, 17:53:42)  [GCC 7.5.0] (64-bit runtime)                                                                                     Python platform: Linux-5.4.0-80-generic-x86_64-with-glibc2.27                                                                                                             Is CUDA available: False
CUDA runtime version: 11.2.152
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti
GPU 4: GeForce RTX 2080 Ti
GPU 5: GeForce RTX 2080 Ti
GPU 6: GeForce RTX 2080 Ti
GPU 7: GeForce RTX 2080 Ti
GPU 8: GeForce RTX 2080 Ti
GPU 9: GeForce RTX 2080 Ti

Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
```

## Additional context

Likely related to [this](https://github.com/pytorch/pytorch/issues/62967) issue, but unresolved with the latest nightly (12/08/21). Forcing the input to be contiguous fixes the problem.


cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @VitalyFedyunin @jamesr66a @ppwwyyxx

**Comment by xwang233** [comment]
https://github.com/pytorch/pytorch/pull/62773 is likely related. cc @eqy

**Comment by eqy** [comment]
Will see if I can bisect this; my understanding was that the contiguous check would have prevented this issue:
#62839

**Comment by eqy** [comment]
I can repro this in NGC 21.07 (which was frozen before #62773 and #62389 so the plot thickens...)

**Comment by VitalyFedyunin** [comment]
Looks like regression to me, need to be fixed before 1.10 cut (should we add new label / project?)

**Comment by eqy** [comment]
I've bisected this to f3218568 (#59286).

**Comment by VitalyFedyunin** [comment]
Thanks, @eqy I will ask the PR author to look into it.

**Comment by mingfeima** [comment]
WIP!

**Comment by mingfeima** [comment]
Update the memory format dispatch logic at #63392 to match 1.9 behavior.

The issue is brought up due to the given input has some sort of ambiguous, the strides are {6, 1, 2, 1}. So it is considered to be CL-contiguous but it has a `suggest_memory_format` of `MemoryFormat::Contiguous`.

1.9 dispatch logic is like:
* input.is_contigous() - go NCHW path
* input.is_contigous(MemoryFormat::ChannelsLast) - go NHWC path

while 1.10 dispatch logic is based on suggest memory format:
On contiguous input (either contiguous or CL-contiguous)
* at::MemoryFormat::Contigous - go NCHW path
* at::MemoryFormat::ChannelsLast - go NHWC path

So it went wrong.
